File: libs/clojure-openapi-codegen/test/openapi_codegen/parser_test.clj
(ns openapi-codegen.parser-test
  (:require [clojure.test :refer [deftest testing is]]
            [openapi-codegen.parser :as sut]))

(def sample-yaml
  "openapi: 3.1.0
info:
  title: Test API
  version: 1.0.0
components:
  schemas:
    LoginRequest:
      type: object
      description: Login credentials
      required:
        - username
        - password
      properties:
        username:
          type: string
        password:
          type: string
    UserProfile:
      type: object
      description: User profile details
      required:
        - id
        - email
      properties:
        id:
          type: string
        email:
          type: string
          format: email
        age:
          type: integer
        score:
          type: number
        active:
          type: boolean
        tags:
          type: array
          items:
            type: string
    EmptySchema:
      type: object
      description: Schema with no properties
")

(defn- write-temp-yaml
  "Write yaml-content to a temp file and return its path string."
  [yaml-content]
  (let [f (java.io.File/createTempFile "test-spec" ".yaml")]
    (.deleteOnExit f)
    (spit f yaml-content)
    (.getAbsolutePath f)))

(deftest parse-schemas-returns-sequence-test
  (testing "returns a non-empty sequence for a spec with schemas"
    (let [spec-path (write-temp-yaml sample-yaml)
          schemas   (sut/parse-schemas spec-path)]
      (is (seq schemas))
      (is (= 3 (count schemas))))))

(deftest parse-schemas-schema-names-test
  (testing "each parsed schema has a :name key"
    (let [spec-path (write-temp-yaml sample-yaml)
          schemas   (sut/parse-schemas spec-path)
          names     (set (map :name schemas))]
      (is (contains? names "LoginRequest"))
      (is (contains? names "UserProfile"))
      (is (contains? names "EmptySchema")))))

(deftest parse-schemas-required-fields-test
  (testing "required properties are marked with :required? true"
    (let [spec-path (write-temp-yaml sample-yaml)
          schemas   (sut/parse-schemas spec-path)
          login     (first (filter #(= "LoginRequest" (:name %)) schemas))
          props     (:properties login)
          username  (first (filter #(= "username" (:name %)) props))
          password  (first (filter #(= "password" (:name %)) props))]
      (is (true? (:required? username)))
      (is (true? (:required? password))))))

(deftest parse-schemas-optional-fields-test
  (testing "non-required properties are marked with :required? false"
    (let [spec-path (write-temp-yaml sample-yaml)
          schemas   (sut/parse-schemas spec-path)
          profile   (first (filter #(= "UserProfile" (:name %)) schemas))
          props     (:properties profile)
          age       (first (filter #(= "age" (:name %)) props))]
      (is (false? (:required? age))))))

(deftest parse-schemas-property-types-test
  (testing "property types are extracted correctly"
    (let [spec-path (write-temp-yaml sample-yaml)
          schemas   (sut/parse-schemas spec-path)
          profile   (first (filter #(= "UserProfile" (:name %)) schemas))
          props     (:properties profile)
          find-prop (fn [n] (first (filter #(= n (:name %)) props)))]
      (is (= "string"  (:type (find-prop "id"))))
      (is (= "integer" (:type (find-prop "age"))))
      (is (= "number"  (:type (find-prop "score"))))
      (is (= "boolean" (:type (find-prop "active"))))
      (is (= "array"   (:type (find-prop "tags")))))))

(deftest parse-schemas-array-items-test
  (testing "array properties carry :items sub-schema"
    (let [spec-path (write-temp-yaml sample-yaml)
          schemas   (sut/parse-schemas spec-path)
          profile   (first (filter #(= "UserProfile" (:name %)) schemas))
          tags      (first (filter #(= "tags" (:name %)) (:properties profile)))]
      (is (some? (:items tags)))
      (is (= "string" (get (:items tags) "type"))))))

(deftest parse-schemas-description-test
  (testing "description is extracted from schema definition"
    (let [spec-path (write-temp-yaml sample-yaml)
          schemas   (sut/parse-schemas spec-path)
          login     (first (filter #(= "LoginRequest" (:name %)) schemas))]
      (is (= "Login credentials" (:description login))))))

(deftest parse-schemas-empty-properties-test
  (testing "schema with no properties yields empty :properties sequence"
    (let [spec-path (write-temp-yaml sample-yaml)
          schemas   (sut/parse-schemas spec-path)
          empty-s   (first (filter #(= "EmptySchema" (:name %)) schemas))]
      (is (empty? (:properties empty-s))))))

(deftest parse-schemas-required-set-test
  (testing ":required key is a set of required property name strings"
    (let [spec-path (write-temp-yaml sample-yaml)
          schemas   (sut/parse-schemas spec-path)
          login     (first (filter #(= "LoginRequest" (:name %)) schemas))]
      (is (set? (:required login)))
      (is (= #{"username" "password"} (:required login))))))

(deftest parse-schemas-no-schemas-test
  (testing "returns empty sequence when components.schemas is absent"
    (let [empty-yaml "openapi: 3.1.0\ninfo:\n  title: Empty\n  version: 1.0.0\n"
          spec-path  (write-temp-yaml empty-yaml)
          schemas    (sut/parse-schemas spec-path)]
      (is (empty? schemas)))))

(deftest parse-schemas-format-preserved-test
  (testing "property format (e.g. email, date-time) is extracted"
    (let [spec-path (write-temp-yaml sample-yaml)
          schemas   (sut/parse-schemas spec-path)
          profile   (first (filter #(= "UserProfile" (:name %)) schemas))
          email     (first (filter #(= "email" (:name %)) (:properties profile)))]
      (is (= "email" (:format email))))))

File: libs/elixir-openapi-codegen/test/parser_test.exs
defmodule OpenApiCodegen.ParserTest do
  use ExUnit.Case, async: true

  alias OpenApiCodegen.Parser

  @fixtures_dir Path.expand("fixtures", __DIR__)
  @sample_yaml Path.join(@fixtures_dir, "sample.yaml")

  describe "parse_file/1" do
    test "returns ok with list of schemas from valid spec file" do
      assert {:ok, schemas} = Parser.parse_file(@sample_yaml)
      assert is_list(schemas)
      assert length(schemas) == 3
    end

    test "schemas are sorted by name" do
      assert {:ok, schemas} = Parser.parse_file(@sample_yaml)
      names = Enum.map(schemas, & &1.name)
      assert names == Enum.sort(names)
    end

    test "parses User schema with correct required fields" do
      assert {:ok, schemas} = Parser.parse_file(@sample_yaml)
      user = Enum.find(schemas, &(&1.name == "User"))
      assert user != nil
      assert Enum.sort(user.required) == Enum.sort(["id", "username", "email"])
    end

    test "parses User schema with all properties" do
      assert {:ok, schemas} = Parser.parse_file(@sample_yaml)
      user = Enum.find(schemas, &(&1.name == "User"))
      prop_names = Enum.map(user.properties, & &1.name)
      assert "id" in prop_names
      assert "username" in prop_names
      assert "email" in prop_names
      assert "displayName" in prop_names
      assert "age" in prop_names
      assert "score" in prop_names
      assert "active" in prop_names
      assert "tags" in prop_names
      assert "metadata" in prop_names
    end

    test "properties have correct types" do
      assert {:ok, schemas} = Parser.parse_file(@sample_yaml)
      user = Enum.find(schemas, &(&1.name == "User"))
      by_name = Map.new(user.properties, &{&1.name, &1})

      assert by_name["id"].type == "string"
      assert by_name["age"].type == "integer"
      assert by_name["score"].type == "number"
      assert by_name["active"].type == "boolean"
      assert by_name["tags"].type == "array"
      assert by_name["metadata"].type == "object"
    end

    test "nullable fields are marked correctly" do
      assert {:ok, schemas} = Parser.parse_file(@sample_yaml)
      user = Enum.find(schemas, &(&1.name == "User"))
      by_name = Map.new(user.properties, &{&1.name, &1})

      assert by_name["age"].nullable == true
      assert by_name["score"].nullable == true
      assert by_name["id"].nullable == false
    end

    test "schema with no properties has empty properties list" do
      assert {:ok, schemas} = Parser.parse_file(@sample_yaml)
      empty = Enum.find(schemas, &(&1.name == "EmptySchema"))
      assert empty != nil
      assert empty.properties == []
      assert empty.required == []
    end

    test "returns error for non-existent file" do
      assert {:error, _reason} = Parser.parse_file("/nonexistent/path.yaml")
    end

    test "returns error when components key is missing" do
      no_components = Path.join(@fixtures_dir, "no_components.yaml")
      assert {:error, reason} = Parser.parse_file(no_components)
      assert reason =~ "components"
    end

    test "returns error when schemas key is missing" do
      no_schemas = Path.join(@fixtures_dir, "no_schemas.yaml")
      assert {:error, reason} = Parser.parse_file(no_schemas)
      assert reason =~ "components.schemas"
    end
  end

  describe "parse_string/1" do
    test "parses valid YAML string" do
      yaml = """
      openapi: "3.1.0"
      components:
        schemas:
          Item:
            type: object
            required:
              - id
            properties:
              id:
                type: string
              count:
                type: integer
                nullable: true
      """

      assert {:ok, schemas} = Parser.parse_string(yaml)
      assert length(schemas) == 1
      [item] = schemas
      assert item.name == "Item"
      assert item.required == ["id"]
      assert length(item.properties) == 2
    end

    test "returns error when components missing in string" do
      # YamlElixir parses valid YAML that simply lacks the components key
      yaml = "openapi: '3.1.0'\ninfo:\n  title: Test\n  version: 1.0.0\n"
      assert {:error, _reason} = Parser.parse_string(yaml)
    end

    test "parses schema with $ref property type as object" do
      yaml = """
      openapi: "3.1.0"
      components:
        schemas:
          Parent:
            type: object
            required:
              - child
            properties:
              child:
                $ref: '#/components/schemas/Child'
      """

      assert {:ok, schemas} = Parser.parse_string(yaml)
      [parent] = schemas
      child_prop = Enum.find(parent.properties, &(&1.name == "child"))
      assert child_prop.type == "object"
    end

    test "parses schema with unknown type as string" do
      yaml = """
      openapi: "3.1.0"
      components:
        schemas:
          Strange:
            type: object
            properties:
              field:
                type: exotic_custom_type
      """

      assert {:ok, schemas} = Parser.parse_string(yaml)
      [strange] = schemas
      [field] = strange.properties
      assert field.type == "string"
    end

    test "handles multiple schemas sorted alphabetically" do
      yaml = """
      openapi: "3.1.0"
      components:
        schemas:
          Zebra:
            type: object
            properties:
              name:
                type: string
          Alpha:
            type: object
            properties:
              id:
                type: integer
      """

      assert {:ok, schemas} = Parser.parse_string(yaml)
      names = Enum.map(schemas, & &1.name)
      assert names == ["Alpha", "Zebra"]
    end
  end
end
